src/backend/distributed/metadata/node_metadata.c

```diff
@@ -573,54 +579,173 @@ master_set_node_property(PG_FUNCTION_ARGS)
+
+		multipleTableIntegrationCommandList = list_concat(
+			multipleTableIntegrationCommandList,
+			foreignConstraintCommands);
+	}
+
+	/* construct partitioning hierarchy after all tables are created */
+	foreach_ptr(cacheEntry, propagatedTableList)
+	{
+		Oid relationId = cacheEntry->relationId;
+
+		if (IsTableOwnedByExtension(relationId))
 		{
-			ereport(WARNING, (errmsg("citus.enable_object_propagation is off, not "
-									 "creating distributed objects on worker"),
-							  errdetail("distributed objects are only kept in sync when "
-										"citus.enable_object_propagation is set to on. "
-										"Newly activated nodes will not get these "
-										"objects created")));
+			/* skip partition creation when the Citus table is owned by an extension */
+			continue;
 		}
 
-		if (ReplicateReferenceTablesOnActivate)
+		if (PartitionTable(relationId))
 		{
-			ReplicateAllReferenceTablesToNode(newWorkerNode->workerName,
-											  newWorkerNode->workerPort);
+			char *alterTableAttachPartitionCommands =
+				GenerateAlterTableAttachPartitionCommand(relationId);
+
+			multipleTableIntegrationCommandList = lappend(
+				multipleTableIntegrationCommandList,
+				alterTableAttachPartitionCommands);
 		}
 	}
+
+	multipleTableIntegrationCommandList = lcons(DISABLE_DDL_PROPAGATION,
+												multipleTableIntegrationCommandList);
+	multipleTableIntegrationCommandList = lappend(multipleTableIntegrationCommandList,
+												  ENABLE_DDL_PROPAGATION);
+
+	return multipleTableIntegrationCommandList;
 }
 
 
 /*
- * PropagateNodeWideObjects is called during node activation to propagate any object that
- * should be propagated for every node. These are generally not linked to any distributed
- * object but change system wide behaviour.
+ * MetadataSetupCommandList get the command list to set up the metadata
+ * depending on the distributed object on the given node.
  */
-static void
-PropagateNodeWideObjects(WorkerNode *newWorkerNode)
+static List *
+MetadataSetupCommandList()
+{
+	List *distributedTableList = CitusTableList();
+	List *propagatedTableList = NIL;
+	List *metadataSnapshotCommandList = NIL;
+
+	/* create the list of tables whose metadata will be created */
+	CitusTableCacheEntry *cacheEntry = NULL;
+	foreach_ptr(cacheEntry, distributedTableList)
+	{
+		if (ShouldSyncTableMetadata(cacheEntry->relationId))
+		{
+			propagatedTableList = lappend(propagatedTableList, cacheEntry);
+		}
+	}
+
+	/* after all tables are created, create the metadata */
+	foreach_ptr(cacheEntry, propagatedTableList)
+	{
+		Oid clusteredTableId = cacheEntry->relationId;
+
+		/* add the table metadata command first*/
+		char *metadataCommand = DistributionCreateCommand(cacheEntry);
+		metadataSnapshotCommandList = lappend(metadataSnapshotCommandList,
+											  metadataCommand);
+
+		/* add the truncate trigger command after the table became distributed */
+		if (!IsForeignTable(clusteredTableId))
+		{
+			char *truncateTriggerCreateCommand =
```

Comment:
Can we move truncateTrigger logic to shell table creation part?

Reply:
I've kept it here as we need to create it even the table is owned by extension and it drops the metadata. Let's discuss it further offline.